### PR TITLE
Time indexed forecast params

### PR DESCRIPTION
**Pull request by kkarrancsu**
Updates to power forecasting module to accept vector or scalar inputs.

### CODE REVIEW

**Review by tmellan — COMMENTED** [review]
curious to know what's the extra value of having the if logic in lines 43-47 instead of

`qa_onboarded_power_vec = np.ones(forecast_lenght) * qa_onboard_power`

which I think should return the same?

**Review by tmellan — APPROVED** [review]

**Review by tmellan — COMMENTED** [review]
in the future, if we're sure this is not a bug but rather is a deficiency in the model popping up via the approximations used, we may way want to include a learnable factor to correct the difference

File: mechafil/power.py [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
 
 # --------------------------------------------------------------------------------------
 #  QA Multiplier functions
 # --------------------------------------------------------------------------------------
 def compute_qa_factor(
-    fil_plus_rate: float,
+    fil_plus_rate: Union[np.array, float],
     fil_plus_m: float = 10.0,
     duration_m: Callable = None,
     duration: int = None,
```
Comment (misilva73) [inline mechafil/power.py:14]: If `fil_plus_rate` is an array, then the return of this function will also be an array. You need to update the return type of this function to `Union[np.array, float]`

File: mechafil/power.py [hunk 2]
```diff
@@ -52,7 +56,8 @@ def compute_day_rb_renewed_power(
     day_scheduled_expire_power_vec: np.array,
     renewal_rate_vec: np.array,
 ):
-    day_renewed_power = renewal_rate_vec[-1:] * day_scheduled_expire_power_vec[day_i]
+    #day_renewed_power = renewal_rate_vec[-1:] * day_scheduled_expire_power_vec[day_i]
```
Comment (misilva73) [inline mechafil/power.py:59]: I think we can delete this comment. The new code is actually cleaner

File: mechafil/power.py [hunk 3]
```diff
@@ -66,8 +71,11 @@ def compute_day_qa_renewed_power(
     duration: int = None,
 ):
     qa_factor = compute_qa_factor(fil_plus_rate, fil_plus_m, duration_m, duration)
+    # day_renewed_power = (
+    #     qa_factor * renewal_rate_vec[-1:] * day_rb_scheduled_expire_power_vec[day_i]
+    # )
     day_renewed_power = (
-        qa_factor * renewal_rate_vec[-1:] * day_rb_scheduled_expire_power_vec[day_i]
+        qa_factor * renewal_rate_vec[day_i] * day_rb_scheduled_expire_power_vec[day_i]
```
Comment (misilva73) [inline mechafil/power.py:78]: `fil_plus_rate` can be an array, right? In that case, `qa_factor` will also be an array and thus we should multiply this expression by `qa_factor[day]`.

Also, do not forget to update the the types in the function definition. I also realised and I forgot to add the return type in the def. Can you add it?

**Review by misilva73 — CHANGES_REQUESTED** [review]
@kkarrancsu , great work!! The code is super clear and easy to follow the changes. I left a few comments (most of them are mostly organisational).

I have two general comments:
1. This PR includes the code that is being merged in #2 . We should merge that one first and then check if we get any conflicts with this PR.
2. The notebook you added is a bit hard to follow. Think of the notebooks here as guides or tutorials for people that are not familiar with the code. It would be great if we can clean the notebook a bit by removing commented code and make the outputs of the cells smaller (e.g., we should hide those huge prints from forecasting models).

File: mechafil/data.py [hunk 4]
```diff
@@ -195,6 +197,19 @@ def get_cum_capped_rb_power(date: datetime.date):
     return init_cum_capped_power
 
 
+def get_cum_capped_qa_power(date: datetime.date):
```
[the remark this answers is not in the document]
Comment (tmellan) [reply mechafil/data.py:200]: If using the forecast model, the prints shouldn't be hidden by default. This is essential information on whether the forecasts are reliable.

[the remark this answers is not in the document]
Comment (misilva73) [reply mechafil/data.py:200]: Are you referring to the notebook?

[the remark this answers is not in the document]
Comment (kkarrancsu) [reply mechafil/data.py:200]: I use the function in this notebook [here](https://github.com/protocol/CryptoEconLab/blob/baseline-crossing-updated/notebooks/baseline_crossing/baseline_crossing_vector_inputs_fofr.ipynb): 
It is in a different repo, but I use it to compare the effect of using RBP vs. QAP for miner returns.

[the remark this answers is not in the document]
Comment (misilva73) [reply mechafil/data.py:200]: Got it. we can keep it then!

File: mechafil/data.py [hunk 5]
```diff
@@ -221,3 +236,12 @@ def query_historical_rb_power() -> pd.DataFrame:
     rbp_df["rb_power"] = rbp_df["total_raw_bytes_power"].astype(float)
     rbp_df = rbp_df[["date", "rb_power"]]
     return rbp_df
+
+def query_historical_qa_power() -> pd.DataFrame:
```
[the remark this answers is not in the document]
Comment (kkarrancsu) [reply mechafil/data.py:240]: Based on above resolution, I think this comment is now OBE.

File: mechafil/power.py (hunk 1 above)
Comment (kkarrancsu) [reply mechafil/power.py:14]: good catch. done.

File: mechafil/power.py (hunk 3 above)
Comment (kkarrancsu) [reply mechafil/power.py:78]: addressed in a better way per your recommendations.

File: mechafil/power.py [hunk 6]
```diff
@@ -105,20 +113,33 @@ def compute_day_se_power(
...
     duration: int,
     forecast_lenght: int,
     fil_plus_m: float = 10.0,
     duration_m: Callable = None,
 ) -> Tuple[pd.DataFrame, pd.DataFrame]:
     # Forecast onboards
-    day_rb_onboarded_power = forecast_rb_daily_onboardings(
-        rb_onboard_power, forecast_lenght
-    )
+    if isinstance(renewal_rate, numbers.Number):
```
[the remark this answers is not in the document]
Comment (kkarrancsu) [reply mechafil/power.py:127]: agreed, done.

File: mechafil/power.py [hunk 7]
```diff
@@ -105,20 +113,33 @@ def compute_day_se_power(
...
+        renewal_rate_vec = np.ones(forecast_lenght) * renewal_rate
+    else:
+        assert len(renewal_rate) == forecast_lenght, \
+            "if renewal_rate is provided as a vector, it must be the same length as the forecast length"
+        renewal_rate_vec = renewal_rate
+
+    if isinstance(rb_onboard_power, numbers.Number):
```
[the remark this answers is not in the document]
Comment (kkarrancsu) [reply mechafil/power.py:134]: moved logic into its own function.  I think the assert statement takes care of the case you discuss, where `len(renewal_rate) != forecast_lenght`

File: mechafil/power.py [hunk 8]
```diff
@@ -155,11 +176,13 @@ def forecast_power_stats(
             day_qa_renewed_power,
             duration,
         )
+        fpr = fil_plus_rate if isinstance(fil_plus_rate, numbers.Number) else fil_plus_rate[day_i]
```
[the remark this answers is not in the document]
Comment (kkarrancsu) [reply mechafil/power.py:179]: yep - agreed, done.

**Comment by kkarrancsu** [comment]
> curious to know in `power.py` what's the extra value of having the if logic in lines 43-47 instead of
> 
> `qa_onboarded_power_vec = np.ones(forecast_lenght) * qa_onboard_power`
> 
> which I think should return the same?

You are right, but this made me realize that if `len(qa_onboard_power) != forecast_lenght`, then there'd be a strange error down the line. I addressed this with a new helper function to hopefully make this clearer.

File: mechafil/power.py [hunk 9]
```diff
@@ -1,17 +1,30 @@
...
+import numbers
+
+# --------------------------------------------------------------------------------------
+#  Utility functions
+# --------------------------------------------------------------------------------------
+def scalar_or_vector_to_vector(input_x, expected_len, err_msg=None):
+    if isinstance(input_x, numbers.Number):
+        return np.ones(expected_len) * input_x
+    else:
+        err_msg_out = "vector input does not match expected length!" if err_msg is None else err_msg
```
Comment (misilva73) [inline mechafil/power.py:14]: Why not make "vector input does not match expected length!" the default input of `err_msg`?

File: mechafil/power.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (misilva73) [reply mechafil/power.py:135]: Maybe you didn't push the latest code version, but the `renewal_rate` is not yet using the `scalar_or_vector_to_vector` function

**Review by misilva73 — CHANGES_REQUESTED** [review]
Looks good! Two last bits and we are ready to merge

File: mechafil/power.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kkarrancsu) [reply mechafil/power.py:135]: thanks for the catch - updated.

File: mechafil/power.py (hunk 9 above)
Comment (kkarrancsu) [reply mechafil/power.py:14]: Mainly for cosmetic reasons, since that would be a really long string in the function header? But, open to changing if you think otherwise. Strings are immutable so I don't think there is a concern as there would be if passing in a dictionary.

**Review by misilva73 — APPROVED** [review]
Looks good. You can merge now @kkarrancsu
